### PR TITLE
ASCII-art emulator

### PR DESCRIPTION
**Pull request by rm-hull**

### CODE REVIEW

File: luma/emulator/device.py [hunk 1]
```diff
@@ -6,17 +6,25 @@
 import sys
 import atexit
 import logging
+import string
+import curses
+import collections
+try:
```
Comment (thijstriemstra) [inline luma/emulator/device.py:12]: this should be avoided, we use py27 and newer.

File: luma/emulator/device.py [hunk 2]
```diff
@@ -197,3 +205,116 @@ def contrast(self, value):
...
+    Pseudo-device that acts like a physical display, except that it converts the
+    image to display into an ASCII-art representation and downscales colors to
+    match the xterm-256 color scheme. Supports 24-bit color depth.
+
+    This device takes hold of the terminal window (using curses), and any output
+    for sysout and syserr is captured and stored, and is replayed when the
+    cleanup method is called.
+
+    Loosely based on https://github.com/ajalt/pyasciigen/blob/master/asciigen.py
+    """
```
Comment (thijstriemstra) [inline luma/emulator/device.py]: add 'since' version nr.

File: luma/emulator/device.py (hunk 1 above)
Comment (rm-hull) [reply luma/emulator/device.py:12]: Can you be more specific?
Not sure what you're referring to with 'this'

Comment (thijstriemstra) [reply luma/emulator/device.py:12]: the import workaround. simply `from io import StringIO` is enough, isn't it?

Comment (rm-hull) [reply luma/emulator/device.py:12]: Not exactly sure, but in py2, I think:
* StringIO.StringIO is byte oriented and works with sys.stdout/stderr
* io.StringIO is unicode oriented and doesn't work with sys.stdout/stderr

Comment (thijstriemstra) [reply luma/emulator/device.py:12]: A test would show the difference.

**Review by thijstriemstra — APPROVED** [review]

**Pull request merged**